CS3219-AY2425S1/cs3219-ay2425s1-project-g24
# feat: Communication Nice To Have

**Main Changes**
Integrate a video & audio functionality to replace the existing chat template. This is integrated using PeerJS and their online default server, we set the user Call ID and matched user Call ID to the following:
- `username + "-" + collabId`
- `matchedUsername + "-" + collabId`
(This is to make their ID unique and prevent random users using the same PeerJS server from joining. An alternative is that we can also **host our own PeerJS server as a microservice?**)

The video & audio functionality includes:
- On/Off Webcam
- On/Off Mic
- Call MatchedUser (Either user can initiate the video call)

**User/Activity Flow**
When the user is redirected to the collaboration page, they will be prompted by the browser to allow access for webcam and microphone. They should be able to see themselves. Only after one of the user initiate call then will both user see one another. Each user has control over their own video & audio.

**How to test?**
If you only have a single webcam, you should see that both screens should show the same stream (aka see yourself)
If you happen to have an additional webcam, you can use two browsers, for example arc and chrome, set the default camera of each browser respectively and you should be able to see that the screen display different streams.

**Additional Changes:**
- Change the default code editor to python
- Comment out the `isValidToken` check within frontend's `middleware.ts`

**Bug Fixes:**
- Bug: When a user close his/her browser or tab, the matched user is unaware and stays in the collaboration session
- Fix: When matched user detects that he/she is the only active user in the session, setTimeout for a end session function is triggered (~15s). Within the 15s, if a user reconnects, we abort the end session.
